File: DSA/LinkedList/singlyLinkedList.c
#include <stdio.h>
#include <stdlib.h>
struct node{
    int data;
    struct node* next;
};
//to count number of nodes in a list
int len(struct node* head){
    int count;
    if (head) count=1;else return 0;
    while (head){
        count++;
        head=head->next;
    }return count;
    }
    //we pass the pointer to the pointer of the head node because we are not going to update the head structure we want to update the value of pointer itself
void insertAtFirst(struct node** head,int data){
    //we allocate the memory dynamically so that the new node exist even after the 
    //after the function ends. usually all the memory allocated by the compiler
    //will be released when the function ends.
    struct node* new=(struct node*) malloc(sizeof(struct node));
    new->data=data;
    new->next=*head;//this returns the pointer to the head not the actual structure that the head points
    *head=new;
}
void printlist(struct node* head){
    struct node* current=head;
    while(current){
        printf("%d->",current->data);
        current=current->next;
    }printf("NULL\n");
}
void deletenode(struct node **head,int index);
int main(){
    struct node* head=NULL;
    insertAtFirst(&head,1);
    insertAtFirst(&head,2);
    printlist(head);
    deletenode(&head,1);
    printlist(head);
}
void deletenode(struct node** head,int index){
    int count=0;
    struct node* current=*head;
    while (count<index-1){
        count++;
        current=current->next;
    }
    struct node* delnode=current->next;
    current->next=current->next->next;
    free(delnode);//the free() function clear the data stored in the memory space pointed by the pointer which was passed argument
}

File: DSA/LinkedList/singly_linked_list.py
class Node:
    def __init__(self,data=None,next=None):
        self.next=next
        self.data=data
    def has_next(self):
        return self.next
class llist:
    def __init__(self):
        self.head=None
        self.length=0
    def MA(self,index):
        count=0
        current=self.head
        while count!=index and current:
            current=current.next
            count+=1
        return current
    def insert(self,data,index):
        if index==self.length and index!=0:
            last=self.MA(self.length-1)
            last.next=Node(data)
            self.length+=1
        elif index==0:
            if self.length>=1:
                newnode=Node(data,self.head)
                self.head=newnode
                self.length+=1
            else:
                self.head=Node(data)
                self.length+=1
        else:

            count=0
            current=self.MA(index-1)
            newnode=Node(data,current.next)
            current.next=newnode
    def __str__(self):
        current=self.head
        list=[]
        while current:
            list.append(current.data)
            current=current.next
        return str(list)
    def delete(self,index):
        if index==0:
            self.head=self.head.next
        elif index==self.length:
            last=self.MA(index-1)
            last.next=None
        else:
            prev=self.MA(index-1)
            prev.next=prev.next.next
        self.length-=1
mylist=llist()
mylist.insert(1,0)
mylist.insert(3,1)
mylist.insert(2,1)
print(mylist)
mylist.delete(2)
print(mylist.length)
print(mylist)
#help(map)